src/messaging/ReliableMessageManager.cpp

```diff
@@ -0,0 +1,656 @@
+{
+    for (int i = 0; i < CHIP_CONFIG_RMP_RETRANS_TABLE_SIZE; i++)
+    {
+        if (RetransTable[i].rc == rc)
+        {
+            RetransTable[i].nextRetransTimeTick = 0;
+            break;
+        }
+    }
+}
+
+bool ReliableMessageManager::CheckAndRemRetransTable(ReliableMessageContext * rc, uint32_t ackMsgId)
+{
+    bool res = false;
+
+    for (int i = 0; i < CHIP_CONFIG_RMP_RETRANS_TABLE_SIZE; i++)
+    {
+        if ((RetransTable[i].rc == rc) && RetransTable[i].msgId == ackMsgId)
+        {
+            // Clear the entry from the retransmision table.
+            ClearRetransmitTable(RetransTable[i]);
+
+#if defined(DEBUG)
+            ChipLogProgress(ExchangeManager, "Rxd Ack; Removing MsgId:%08" PRIX32 " from Retrans Table", ackMsgId);
+#endif
+            res = true;
+            break;
+        }
+
+        continue;
+    }
+
+    return res;
+}
+
+/**
+ *  Send the specified entry from the retransmission table.
+ *
+ *  @param[in]    entry                A pointer to a retransmission table entry object that needs to be sent.
+ *
+ *  @return  #CHIP_NO_ERROR On success, else corresponding CHIP_ERROR returned from SendMessage.
+ *
+ */
+CHIP_ERROR ReliableMessageManager::SendFromRetransTable(RetransTableEntry * entry)
+{
+    CHIP_ERROR err              = CHIP_NO_ERROR;
+    ReliableMessageContext * rc = entry->rc;
+
+    // To trigger a call to OnSendError, set the number of transmissions so
+    // that the next call to ExecuteActions will abort this entry,
+    // restart the timer immediately, and ExitNow.
+
+    CHIP_FAULT_INJECT(FaultInjection::kFault_RMPSendError, entry->sendCount = static_cast<uint8_t>(rc->mConfig.mMaxRetrans + 1);
+                      entry->nextRetransTimeTick = 0; StartTimer(); ExitNow());
+
+    if (rc)
+    {
+        // Locally store the start and length;
+        uint8_t * p  = entry->msgBuf->Start();
+        uint32_t len = entry->msgBuf->DataLength();
+
+        // Send the message through
+        uint16_t msgSendFlags = entry->msgSendFlags;
+        SetFlag(msgSendFlags, kChipMessageFlag_RetainBuffer);
+        err = rc->mDelegate.SendMessage(entry->msgBuf, msgSendFlags);
+
+        // Reset the msgBuf start pointer and data length after sending
+        entry->msgBuf->SetStart(p);
+        entry->msgBuf->SetDataLength(static_cast<uint16_t>(len));
+
+        // Update the counters
+        entry->sendCount++;
+    }
+    else
+    {
+        ChipLogError(ExchangeManager, "Table entry invalid");
+    }
+
+    VerifyOrExit(err != CHIP_NO_ERROR, err = CHIP_NO_ERROR);
+
+    // Any error generated during initial sending is evaluated for criticality which would
+    // qualify it to be reportable back to the caller. If it is non-critical then
+    // err is set to CHIP_NO_ERROR.
+    if (IsSendErrorNonCritical(err))
+    {
+        ChipLogError(ExchangeManager, "Non-crit err %ld sending CHIP MsgId:%08" PRIX32 " from retrans table", long(err),
```

Comment:
Should we create a separate category of ReliableMessaging instead of ExchangeManager since this is now removed from ExchangeManager?